### PR TITLE
Add an option to disable diagnostics

### PR DESCRIPTION
**Pull request by popzxc**
As far as I know, currently it's not possible to disable a selected type of diagnostics provided by `rust-analyzer`.

This causes an inconvenient situation with a false-positive warnings: you either have to disable all the diagnostics, or you have to ignore these warnings.

There are some open issues related to this problem, e.g.: https://github.com/rust-analyzer/rust-analyzer/issues/5412, https://github.com/rust-analyzer/rust-analyzer/issues/5502

This PR attempts to make it possible to selectively disable some diagnostics on per-project basis.

### CODE REVIEW

File: crates/ra_hir_expand/src/diagnostics.rs [hunk 1]
```diff
@@ -21,6 +21,7 @@ use ra_syntax::{SyntaxNode, SyntaxNodePtr};
 use crate::{db::AstDatabase, InFile};
 
 pub trait Diagnostic: Any + Send + Sync + fmt::Debug + 'static {
+    fn name(&self) -> String;
```
Comment (jonas-schievink) [inline crates/ra_hir_expand/src/diagnostics.rs:24]:
```suggestion
    fn name(&self) -> &'static str;
```

File: crates/ra_ide/src/diagnostics.rs [hunk 2]
```diff
@@ -362,6 +383,54 @@ mod tests {
         assert_eq!(diagnostics.len(), 0, "unexpected diagnostics:\n{:#?}", diagnostics);
     }
 
+    /// Takes a multi-file input fixture with annotated cursor position and the list of disabled diagnostics,
+    /// and checks that provided diagnostics aren't spawned during analysis.
+    fn check_disabled_diagnostics(
+        ra_fixture: &str,
+        disabled_diagnostics: impl IntoIterator<Item = String>,
+    ) {
+        let disabled_diagnostics: std::collections::HashSet<_> =
```
Comment (jonas-schievink) [inline crates/ra_ide/src/diagnostics.rs:392]: Suggest importing `HashSet` instead

```suggestion
        let disabled_diagnostics: HashSet<_> =
```

File: crates/ra_ide/src/diagnostics.rs [hunk 3]
```diff
@@ -362,6 +383,54 @@ mod tests {
         assert_eq!(diagnostics.len(), 0, "unexpected diagnostics:\n{:#?}", diagnostics);
     }
 
+    /// Takes a multi-file input fixture with annotated cursor position and the list of disabled diagnostics,
+    /// and checks that provided diagnostics aren't spawned during analysis.
+    fn check_disabled_diagnostics(
+        ra_fixture: &str,
+        disabled_diagnostics: impl IntoIterator<Item = String>,
```
Comment (jonas-schievink) [inline crates/ra_ide/src/diagnostics.rs:390]:
```suggestion
        disabled_diagnostics: &[&str],
```

**Comment by popzxc** [comment]
@jonas-schievink 

> Is there a way to configure this in VS Code with this PR?

I don't think so. I initially thought that it will be two different features, and thus have to be implemented separately. If you'll give me a clue where the extension provides a configuration for a server, I can implement it (either as a part of this PR, or as a different PR).

File: crates/ra_ide/src/diagnostics.rs [hunk 4]
```diff
@@ -871,6 +869,6 @@ struct Foo {
 
     #[test]
     fn test_disabled_diagnostics() {
-        check_disabled_diagnostics(r#"mod foo;"#, vec!["unresolved-module".to_string()]);
+        check_disabled_diagnostics(r#"mod foo;"#, &vec!["unresolved-module"]);
```
Comment (jonas-schievink) [inline crates/ra_ide/src/diagnostics.rs:872]:
```suggestion
        check_disabled_diagnostics(r#"mod foo;"#, &["unresolved-module"]);
```

**Comment by jonas-schievink** [comment]
You need to add the new setting to editors/code/package.json, and that should be it

File: editors/code/package.json [hunk 5]
```diff
@@ -904,4 +913,4 @@
             ]
         }
     }
-}
+}
```
Comment (popzxc) [inline editors/code/package.json:916]: For some reason, VS Code auto-formatter removes the trailing newline from `json` files specifically. I can manually restore this newline, but, I guess, anyone who use VS Code with enabled on-save formatting, and will be editing this file, will face the same issue.

Comment (Coder-256) [reply editors/code/package.json:916]: You need to set `"files.insertFinalNewline": true`.

Comment (popzxc) [reply editors/code/package.json]: I see, thanks!

**Comment by popzxc** [comment]
@jonas-schievink Is there anything else that has to be done within this PR?

File: crates/ide/src/lib.rs [hunk 6]
```diff
@@ -99,8 +99,15 @@ pub use text_edit::{Indel, TextEdit};
 
 pub type Cancelable<T> = Result<T, Canceled>;
 
+/// Configuration parameters for the analysis run.
+#[derive(Debug, Default, Clone)]
+pub struct AnalysisConfig {
+    pub disabled_diagnostics: HashSet<String>,
+}
```
Comment (jonas-schievink) [inline crates/ide/src/lib.rs:106]: @matklad is it okay to make this config structure part of `AnalysisHost` or should it be handled differently?

Comment (matklad) [reply crates/ide/src/lib.rs:106]: Good question! Given that we only do surface filtering at the moment, I think it's better to just pass it in as an argument to `Analysis::diagnostics`.

If, in the future, we decide to disable diagnostics on a more fundamental level (ie, avoid computing them at all), I think we'll have to make this an input of the database. 

But for now, adding an argument seems like a better approach.

Comment (popzxc) [reply crates/ide/src/lib.rs:106]: Sure, will do that.

Comment (popzxc) [reply crates/ide/src/lib.rs]: Done.

**Pull request merged**